File: small_scale_channel_tracking/mode_selection_by_SE/SE_vs_rT_plot.py
import os
import scipy.io as sio
import matplotlib.pyplot as plt

# --- 1. Path Setup ---
script_dir = os.path.dirname(os.path.abspath(__file__))
result_dir = os.path.join(script_dir, "SE_vs_rT_result")

FREQ_GHZ = 38
mat_filename = f"SE_vs_rT_{FREQ_GHZ}GHz_data.mat"
mat_filepath = os.path.join(result_dir, mat_filename)

if not os.path.exists(mat_filepath):
    print(f"Error: Could not find the data file at {mat_filepath}")
    exit()

# --- 2. Load Data ---
print(f"Loading data from {mat_filename}...")
data = sio.loadmat(mat_filepath)

SNR_LIST = data['SNR_LIST'].flatten()
RT_RANGE = data['RT_RANGE'].flatten()

# --- 3. Plotting Loop for Each SNR ---
for snr in SNR_LIST:
    print(f"\nGenerating plots for SNR = {snr} dB...")
    snr_str = f"n{abs(snr)}" if snr < 0 else f"p{snr}"
    
    baseline = data[f'baseline_snr_{snr_str}'].flatten()[0]
    res_ideal = data[f'ideal_snr_{snr_str}'].flatten()
    res_hybrid_E = data[f'hybrid_snr_{snr_str}'].flatten()

    # --- A. Generate the Line Plot (IEEE Compliant) ---
    # 3.5 inches is the standard width for an IEEE single-column figure
    plt.figure(figsize=(3.5, 2.6))
    
    plt.plot(RT_RANGE, res_ideal, '-', color='b', linewidth=1.2, 
             label='Modal Fully Digital')
    plt.plot(RT_RANGE, res_hybrid_E, '--o', color='g', linewidth=1.2, markersize=3, 
             label='Modal Hybrid w/ E')

    # Plot the full-digital baseline
    plt.axhline(y=baseline, color='k', linestyle=':', linewidth=1.5, label='Port Bound')

    # IEEE requires smaller font sizes (8pt to 9pt)
    plt.xlabel('Number of Modes ($r_T$)', fontsize=9)
    plt.ylabel('Spectral Efficiency (bps/Hz)', fontsize=9)
    
    plt.grid(True, linestyle=':', alpha=0.6)
    
    # Adjust tick font sizes for print readability
    plt.tick_params(axis='both', which='major', labelsize=8)
    
    # Compact legend without a bulky frame
    plt.legend(loc='lower right', fontsize=7, frameon=False)
    
    plot_path = os.path.join(result_dir, f'SE_vs_rT_curve_{FREQ_GHZ}GHz_SNR_{snr_str}dB.png')
    
    # dpi=300 is the strict IEEE minimum for publication quality
    plt.savefig(plot_path, bbox_inches='tight', dpi=300)
    plt.close()

    # --- B. Generate the Efficiency Table ---
    table_data = []
    for idx, rt in enumerate(RT_RANGE):
        val_i = res_ideal[idx]
        val_e = res_hybrid_E[idx]
        
        eff_i = (val_i / baseline) * 100
        eff_e = (val_e / baseline) * 100
        
        row = [
            f"r_T = {rt}",
            f"{val_i:.2f} ({eff_i:.1f}%)",
            f"{val_e:.2f} ({eff_e:.1f}%)"
        ]
        table_data.append(row)

    col_labels = ['Mode Config', 'Ideal Mode', 'Hybrid w/ E']
    col_colors = ["#f2f2f2", "#d9ead3", "#cfe2f3"]

    fig, ax = plt.subplots(figsize=(6, 12)) 
    ax.axis('off')

    the_table = ax.table(cellText=table_data, 
                         colLabels=col_labels, 
                         colColours=col_colors,
                         cellLoc='center', 
                         loc='center')

    the_table.auto_set_font_size(False)
    the_table.set_fontsize(10)

    for position, cell in the_table.get_celld().items():
        cell.set_height(0.025) 
        if position[0] == 0: 
            cell.set_text_props(weight='bold')
            cell.set_height(0.035)

    plt.title(f'Efficiency Analysis (SNR={snr}dB, {FREQ_GHZ}GHz Imperfect CSI)\nBaseline: Port Domain Bound', 
              y=0.97, fontweight='bold', fontsize=12)

    table_path = os.path.join(result_dir, f'SE_vs_rT_Table_{FREQ_GHZ}GHz_SNR_{snr_str}dB.png')
    plt.savefig(table_path, bbox_inches='tight', dpi=300)
    plt.close()

print(f"\nSuccess! All IEEE plots and tables have been saved to:\n{result_dir}")

File: small_scale_channel_tracking/mode_selection_by_SE/SE_vs_rT_plot.m
clc; clear; close all;

%% --- 1. Path Setup ---
% Get the directory where this script is saved
script_dir = fileparts(mfilename('fullpath'));
result_dir = fullfile(script_dir, 'SE_vs_rT_result');

freq_ghz = 12;
mat_filename = sprintf('SE_vs_rT_%dGHz_data.mat', freq_ghz);
mat_filepath = fullfile(result_dir, mat_filename);

if ~exist(mat_filepath, 'file')
    error('Error: Could not find the data file at:\n%s\nRun your Python simulation first.', mat_filepath);
end

%% --- 2. Load Data ---
fprintf('Loading data from %s...\n', mat_filename);
data = load(mat_filepath);

% Extract core vectors
snr_list = data.SNR_LIST;
rt_range = data.RT_RANGE;

%% --- 3. Plotting Loop for Each SNR ---
for idx = 1:length(snr_list)
    snr = snr_list(idx);
    fprintf('\nGenerating editable figure for SNR = %d dB...\n', snr);
    
    % Reconstruct the dynamic naming string used by Python (e.g., 'n4', 'p0', 'p4')
    if snr < 0
        snr_str = sprintf('n%d', abs(snr));
    else
        snr_str = sprintf('p%d', snr);
    end
    
    % Dynamically fetch the variables from the loaded structure
    baseline = data.(sprintf('baseline_snr_%s', snr_str));
    res_ideal = data.(sprintf('ideal_snr_%s', snr_str));
    res_hybrid = data.(sprintf('hybrid_snr_%s', snr_str));
    
    %% --- 4. Generate the Figure (IEEE Clean Style by default) ---
    % Create a clean, standard aspect-ratio figure canvas
    h_fig = figure('Units', 'inches', 'Position', [2, 2, 5, 3.8]);
    hold on;
    
    % Plot curves with clean styling (thicker lines for easy viewing)
    plot(rt_range, res_ideal, 'b-', 'LineWidth', 1.8, 'DisplayName', 'Modal Fully Digital');
    plot(rt_range, res_hybrid, 'g--o', 'LineWidth', 1.5, 'MarkerSize', 4, 'MarkerFaceColor', 'g', 'DisplayName', 'Modal Hybrid w/ E');
    
    % Plot the Port full-digital ceiling limit
    yline(baseline, 'k:', 'LineWidth', 1.8, 'DisplayName', 'Port Bound');
    
    %% --- 5. Figure Details and Labeling ---
    xlabel('Number of Modes (r_T)', 'FontSize', 11, 'FontName', 'Helvetica');
    ylabel('Spectral Efficiency (bps/Hz)', 'FontSize', 11, 'FontName', 'Helvetica');
    title(sprintf('SE Analysis (SNR = %d dB, %d GHz Imperfect CSI)', snr, freq_ghz), 'FontSize', 11, 'FontWeight', 'bold');
    
    grid on;
    set(gca, 'GridLineStyle', ':', 'GridAlpha', 0.6);
    set(gca, 'FontSize', 10, 'FontName', 'Helvetica'); % Axis tick font sizes
    
    % Add clean legend without a bulky border box
    legend('Location', 'southeast', 'FontSize', 9, 'Box', 'off');
    
    hold off;
    
    %% --- 6. Save Simultaneously as Editable .fig and Publication-Quality .png ---
    % 1. Save the editable MATLAB layout file
    save_filename_fig = sprintf('SE_vs_rT_SNR_%sdB.fig', snr_str);
    save_filepath_fig = fullfile(result_dir, save_filename_fig);
    saveas(h_fig, save_filepath_fig);
    
    % 2. Save the raster image file (300 DPI minimum for IEEE print layout)
    save_filename_png = sprintf('SE_vs_rT_SNR_%sdB.png', snr_str);
    save_filepath_png = fullfile(result_dir, save_filename_png);
    exportgraphics(h_fig, save_filepath_png, 'Resolution', 300);
    
    % Close the active canvas to clear memory
    close(h_fig); 
    
    fprintf('Saved editable fig: %s\n', save_filename_fig);
    fprintf('Saved viewing png:  %s\n', save_filename_png);
end

fprintf('\nSuccess! All figures are safely saved and fully editable inside MATLAB.\n');